prov/efa/src/rxr/rxr_pkt_type_req.c

```diff
@@ -0,0 +1,603 @@
+					   data, data_size);
+	if (!bytes_left) {
+		/*
+		 * rxr_cq_handle_rx_completion() releases pkt_entry, thus
+		 * we do not release it here.
+		 */
+		rxr_cq_handle_rx_completion(ep, pkt_entry, rx_entry);
+		rxr_msg_multi_recv_free_posted_entry(ep, rx_entry);
+		rxr_release_rx_entry(ep, rx_entry);
+		ret = 0;
+	} else {
+		/*
+		 * long message protocol
+		 */
+#if ENABLE_DEBUG
+		dlist_insert_tail(&rx_entry->rx_pending_entry, &ep->rx_pending_list);
+		ep->rx_pending++;
+#endif
+		rx_entry->state = RXR_RX_RECV;
+		rx_entry->tx_id = rxr_get_long_rtm_base_hdr(pkt_entry->pkt)->tx_id;
+		rx_entry->credit_request = rxr_env.tx_min_credits;
+		ret = rxr_pkt_post_ctrl_or_queue(ep, RXR_RX_ENTRY, rx_entry, RXR_CTS_PKT, 0);
+		rxr_pkt_entry_release_rx(ep, pkt_entry);
+	}
+
+	return ret;
+}
+
+ssize_t rxr_pkt_proc_msgrtm(struct rxr_ep *ep,
+			    struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_rx_entry *rx_entry;
+
+	rx_entry = rxr_pkt_get_msgrtm_rx_entry(ep, &pkt_entry);
+	if (OFI_UNLIKELY(!rx_entry))
+		return -FI_ENOBUFS;
+
+	if (rx_entry->state == RXR_RX_MATCHED)
+		return rxr_pkt_proc_matched_rtm(ep, rx_entry, pkt_entry);
+
+	return 0;
+}
+
+ssize_t rxr_pkt_proc_tagrtm(struct rxr_ep *ep,
+			    struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_rx_entry *rx_entry;
+
+	rx_entry = rxr_pkt_get_tagrtm_rx_entry(ep, &pkt_entry);
+	if (OFI_UNLIKELY(!rx_entry))
+		return -FI_ENOBUFS;
+
+	if (rx_entry->state == RXR_RX_MATCHED)
+		return rxr_pkt_proc_matched_rtm(ep, rx_entry, pkt_entry);
+
+	return 0;
+}
+
+/*
+ * proc() functions called by rxr_pkt_handle_recv_completion()
+ */
+ssize_t rxr_pkt_proc_rtm(struct rxr_ep *ep,
+			 struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_base_hdr *base_hdr;
+
+	base_hdr = rxr_get_base_hdr(pkt_entry->pkt);
+	assert(base_hdr->type >= RXR_BASELINE_REQ_PKT_BEGIN);
+
+	switch (base_hdr->type) {
+	case RXR_EAGER_MSGRTM_PKT:
+	case RXR_LONG_MSGRTM_PKT:
+		return rxr_pkt_proc_msgrtm(ep, pkt_entry);
+	case RXR_EAGER_TAGRTM_PKT:
+	case RXR_LONG_TAGRTM_PKT:
+		return rxr_pkt_proc_tagrtm(ep, pkt_entry);
+	default:
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL,
+			"Unknown packet type ID: %d\n",
+		       base_hdr->type);
+		if (rxr_cq_handle_cq_error(ep, -FI_EINVAL))
+			assert(0 && "failed to write err cq entry");
+	}
+
+	return -FI_EINVAL;
+}
+
+void rxr_pkt_handle_rtm_recv(struct rxr_ep *ep,
+			     struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_peer *peer;
+	int ret, msg_id;
+
+	assert(rxr_get_base_hdr(pkt_entry->pkt)->type >= RXR_BASELINE_REQ_PKT_BEGIN);
+
+	peer = rxr_ep_get_peer(ep, pkt_entry->addr);
+	assert(peer);
+
+	if (ep->core_caps & FI_SOURCE)
+		rxr_pkt_post_connack(ep, peer, pkt_entry->addr);
+
+	if (rxr_env.enable_shm_transfer && peer->is_local) {
+		/* no need to reorder msg for shm_ep
+		 * rxr_pkt_proc_rtm will write error cq entry if needed
+		 */
+		rxr_pkt_proc_rtm(ep, pkt_entry);
+		return;
+	}
+
+	msg_id = rxr_pkt_rtm_msg_id(pkt_entry);
+	if (rxr_need_sas_ordering(ep)) {
```

Comment:
While you're here might as well refactor this to 
```
if (!rxr_need_sas_ordering(ep))
        goto proc_msg;
```

Reply:
Not sure about that. I usually prefer positive condition: if something, then something.

Reply:
I see your point now. Doing this would allow us merge two branches, and eliminate one if.

Reply:
Yep, normally I prefer that too, but in this case it will help with indentation and isn't too much harder to parse.

Reply:
I have a patch, but since this an unrelated change. I will open a separate PR.

Reply:
I don't think it's unrelated, but if you want to do a separate PR okay.

Reply:
I did not notice this a brand new file, included the change in new revision.